doc/grid_ufuncs.rst

```diff
@@ -0,0 +1,443 @@
+
+For each variable, the signature tells us the ``xgcm.Axis`` positions we require that variable to have,
+both before and after our grid ufunc is applied.
+This information is encoded in the format ``axis_name:position``.
+Each variable can be operated on along multiple axes, which are separated by a comma, e.g. ``(ax1:left, ax2:right)``.
+
+The axis names used in the signature are dummy names: they do not have to be the same as the axis names used in your ``Grid`` object.
+This allows you to write a grid ufunc that can accept axis with any name.
+Therefore the signature ``"(ax1:center)->(ax1:left)"`` means all of:
+
+`"This function accepts one data variable and applies an operation over one core dimension.
+The input data lies on the center grid positions of the singular axis along which we will choose to apply the function.
+After performing its numerical operation the single return value from this function will have been shifted onto the left-hand grid positions of the same axis."`
+
+To give you an idea of how we might use grid ufuncs here is a table of possible grid ufuncs and their corresponding signatures:
+
+.. list-table:: Example signatures for different grid ufuncs
+   :widths: 25 25 50
+   :header-rows: 1
+
+   * - Name
+     - Signature
+     - Description
+   * - ``diff_center_to_left(a)``
+     - ``"(X:center)->(X:left)"``
+     - Backward difference
+   * - ``interp_center_to_left(a)``
+     - ``"(X:center)->(X:left)"``
+     - Backward interpolation
+   * - ``diff_left_to_center(a)``
+     - ``"(X:left)->(X:center)"``
+     - Forward difference
+   * - ``interp_left_to_center(a)``
+     - ``"(X:left)->(X:center)"``
+     - Forward interpolation
+   * - ``diff_center_to_center(a)``
+     - ``"(X:center)->(X:center)"``
+     - Second order central difference
+   * - ``mean_depth(w)``
+     - ``"(depth:center)->()"``
+     - Reduction
+   * - ``inner_product_left_right(a, b)``
+     - ``"(X:left),(X:right)->()"``
+     - Reduction of two variables
+   * - ``u_grid_vorticity(u, v)``
+     - ``"(lon:left,lat:center),``
+       ``(lon:center,lat:left)``
+       ``->(lon:left,lat:left)"``
+     - Complex example calculating vorticity on an Arakawa U-grid in 2D
+
+.. note::
+
+    Remember the axis names in the signature are dummy names - in the example above you could apply ``mean_depth`` along
+    an axis not called ``"depth"`` if you wish.
+
+    The ``axis`` argument is the one which must correspond to an ``xgcm.Axis`` of the grid.
+    Therefore applying a grid ufunc with signature ``"(X:center)->()"`` or ``"(depth:center)->()"`` along ``axis='X'`` will
+    yield identical results in both cases.
+
+
+Defining New Grid Ufuncs
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Lets imagine we have a numpy function which does forward differencing along one dimension, with an implicit periodic boundary condition.
+
+.. ipython:: python
+
+    import numpy as np
+
+    def diff_forward(a):
+        return a - np.roll(a, -1, axis=-1)
```

Comment:
I find the use of roll here a little confusing, because it brings up boundary conditions. We haven't really introduced how boundary conditions will be handled yet.

Reply:
Actually, it seems fine after reading the next section.